File: javascript/clock/clock.js
//@ts-check

export class Clock {
  constructor(hours = 0, minutes = 0) {
    hours = (hours + Math.floor(minutes / 60)) % 24;
    minutes = minutes % 60;
    /** @type {number} */
    this.hours = hours < 0 ? 24 + hours : hours;
    /** @type {number} */
    this.minutes = minutes < 0 ? 60 + minutes : minutes;
  }

  /**
   * @param {number} hours
   * @param {number} minutes
   * @returns {Clock}
   */
  static at(hours, minutes) {
    return new Clock(hours, minutes);
  }

  /**
   * @param {Clock} clock
   * @returns {boolean}
   */
  equals(clock) {
    if (!(clock instanceof Clock)) {
      throw new TypeError('Not a valid clock');
    }
    return this.toString() == clock.toString();
  }

  /**
   * @param {number} minutes
   * @returns {Clock}
   */
  minus(minutes = 0) {
    return Clock.at(this.hours, this.minutes - minutes);
  }

  /**
   * @param {number} minutes
   * @returns {Clock}
   */
  plus(minutes = 0) {
    return Clock.at(this.hours, this.minutes + minutes);
  }

  toString() {
    return `${this.hours.toString().padStart(2, '0')}:${this.minutes.toString().padStart(2, '0')}`;
  }
}

File: go/clock/clock.go
package clock

import (
	"fmt"
	"math"
)

// Clock is a clock
type Clock struct {
	hours   int
	minutes int
}

// New returns a new Clock
func New(hours, minutes int) Clock {
	hours += int(math.Floor(float64(minutes/60))) % 24
	minutes = minutes % 60

	if hours < 0 {
		hours += 24
	}

	if minutes < 0 {
		minutes += 60
	}

	return Clock{
		hours,
		minutes,
	}
}

// Subtract subtracts minutes from a clock
func (clock Clock) Subtract(minutes int) Clock {
	hours := minutes / 60
	minutes = minutes % 60
	clock.hours -= hours
	clock.minutes -= minutes
	return clock
}

// Add adds minutes to a clock
func (clock Clock) Add(minutes int) Clock {
	hours := minutes / 60
	minutes = minutes % 60
	clock.hours += hours
	clock.minutes += minutes
	return clock
}

// String returns a string representation of the clock
func (clock Clock) String() string {
	return fmt.Sprintf("%02d:%02d", clock.hours, clock.minutes)
}
